src/wp-includes/formatting.php

```diff
@@ -2146,17 +2274,21 @@ function sanitize_file_name( $filename ) {
 function sanitize_user( $username, $strict = false ) {
 	$raw_username = $username;
 	$username     = wp_strip_all_tags( $username );
-	$username     = remove_accents( $username );
 	// Remove percent-encoded characters.
-	$username = preg_replace( '|%([a-fA-F0-9][a-fA-F0-9])|', '', $username );
+	$username = urldecode( $username );
 	// Remove HTML entities.
 	$username = preg_replace( '/&.+?;/', '', $username );
 
-	// If strict, reduce to ASCII for max portability.
-	if ( $strict ) {
+	// If mixing different scripts, remove all but ASCII.
+	if ( ! uses_single_unicode_script( $username ) ) {
 		$username = preg_replace( '|[^a-z0-9 _.\-@]|i', '', $username );
 	}
 
+	// If strict, remove reduce to letters and numbers.
+	if ( $strict ) {
+		$username = preg_replace( '|[^a-z0-9 _.\-@\p{L}\p{N}]|iu', '', $username );
```

Comment:
there are many existing cases missing this, but we should check for UTF-8 PCRE support before calling it otherwise this is guaranteed to set `$username = false`

```php
$username = _wp_can_use_pcre_u()
	? preg_replace( '|…|iu', '', $username )
	: preg_replace( '|[^a-z0-9 _.\-@]|', '', $username );
```

though it would be possible still to fallback to a manual replacement method if we care that much.

this all highlights how much we could carefully revise all of the assumptions in these functions and think through the implications of what we allow or not

Reply:
I didn't bother with any test since existing code assumed that /u is supported. [Here](https://github.com/WordPress/wordpress-develop/commit/943fa44db03#diff-5ff39f4c64b9bebb085a89f0c7e6242b329f1ac0a10855f42dc0223c7887804eL502) is a commit from 2019.

Reply:
Well, OK.

I see you even added a way to mock absence. Will push something suitable.